plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/conversions/RemoveUnnecessaryParenthesesConversion.kt

```diff
@@ -0,0 +1,213 @@
+            return true
+        }
+
+        // Some parent nodes always surround their child with unambiguous whitespace/keywords/parentheses/commands/etc. and have only one
+        // child node that can possibly be surrounded by parentheses
+        if (parent is JKDelegationConstructorCall ||
+            parent is JKKtWhenCase ||
+            parent is JKReturnStatement ||
+            parent is JKArgument
+        ) {
+            return false
+        }
+
+        // Keep parentheses on the receiver in expressions like `(str1 + str2).length`
+        if ((parent is JKQualifiedExpression && parent.receiver == expression) ||
+            (parent is JKAssignmentChainAlsoLink && parent.receiver == expression)
+        ) {
+            return !isAtomic(innerExpression)
+        }
+
+        // Conditions in if-else/while/do-while expressions have their own parentheses (e.g. the `(isEnabled)` in `if (isEnabled) return`)
+        if ((parent is JKIfElseStatement && parent.condition === expression) ||
+            (parent is JKWhileStatement && parent.condition === expression) ||
+            (parent is JKDoWhileStatement && parent.condition === expression)
+        ) {
+            return false
+        }
+
+        if (parent is JKForInStatement && parent.iterationExpression == expression) return false
+
+        if (parent is JKIsExpression && parent.expression == expression) return false
+
+        // No need for parentheses if the expression is a standalone statement living in a block
+        val grandparent = parent.parent
+        if (parent is JKExpressionStatement && (grandparent is JKLambdaExpression || grandparent is JKBlockImpl)) {
+            return false
+        }
+
+        // if parenthesized expression is the initializer, e.g. `val x = (doThing())`
+        if ((parent is JKLocalVariable && parent.initializer == expression) ||
+            (parent is JKField && parent.initializer == expression) ||
+            (parent is JKKtAssignmentStatement && parent.expression == expression)
+        ) {
+            return false
+        }
+
+
+        if (parent is JKPrefixExpression || parent is JKPostfixExpression) {
+            return !isAtomic(innerExpression)
+        }
+
+        // remove parentheses if doing so doesn't affect the order of operations
+        val innerPriority = getPriority(innerExpression)
+        val parentPriority = getPriority(parent)
+        if (innerPriority == parentPriority) {
+            if (parent !is JKBinaryExpression) return false
+            if (innerExpression is JKBinaryExpression &&
+                (innerExpression.operator.token == JKOperatorToken.ANDAND ||
+                        innerExpression.operator.token == JKOperatorToken.OROR)
+            ) {
+                return false
+            }
+            return parent.right == expression
+        }
+
+        return innerPriority < parentPriority
+    }
+
+    private fun isAtomic(expression: JKExpression): Boolean {
```

Comment:
Is this the negation of checks in `org.jetbrains.kotlin.nj2k.ExpressionsKt#parenthesizeIfCompoundExpression`?

Reply:
Not quite: we can't replace one with the negation of the other because there are a bunch of expression types that are neither compound nor atomic (e.g. JKPrefixExpression)